SNMPv2 Traps

florimondmanca opened the pull request:
### What does this PR do?

<!-- A brief description of the change being made with this pull request. -->

Add minimal ("MVP") support for receiving SNMP traps (by listening on UDP sockets) and forwarding them as logs to Datadog.

Adds support for the following:

- Configure via `datadog.yaml`
- Listen for incoming trap packets
  - SNMPv2c only
  - Single listener (single port) with support for multiple community strings. (We can add multi-listener later if/when tackling v3, but it's not required for v2 since we can multiplex packets with different community strings on the same connection.)
  - Wrong client credentials are handled gracefully
- Format and forward trap packets to the logs intake - new logs `input`.
  - Trap packets are parsed and relevant "raw-looking" data is formatted as JSON log messages (no need for a log parser). (Not obvious if this is the best option, but we need to ship something and get feedback on the message format.)
- Update Agent status to report new log source - see https://github.com/DataDog/datadog-agent/pull/5470#issuecomment-654307493
- Update Agent status to expose new expvars in a new "SNMP Traps" status section - see https://github.com/DataDog/datadog-agent/pull/5470#issuecomment-654343061

### Motivation

<!-- What inspired you to submit this pull request? -->

### Additional Notes

<!-- Anything else we should know when reviewing? -->

Out of scope for this PR - may/will be done later:

- SNMPv3 support
  - v3 support in GoSNMP is unstable (see note in the [GoSNMP docs](https://godoc.org/github.com/soniah/gosnmp#NewTrapListener)) and there are bugs, see https://github.com/DataDog/datadog-agent/pull/5470#issuecomment-633641654.
- Trap filtering logic (for now we accept all traps, we'll see if this poses a problem IRL)
- Human-readable OID names (MIB parsing)
- Kubernetes/Helm configuration (will update templates in the appropriate repos later)

### Describe your test plan

<!-- Write there any instructions and details you may have to test your PR. -->

- [x] Unit tests for configuration logic
- [x] Integration tests for trap listeners logic (`TrapServer`):
  - [x] Config file reading
  - [x] Send/receive v2 traps
  - [x] Verify system performs well under high concurrency
- [x] Test log collection
- [x] Test log formatting
- [ ] Add perf benchmark test, see https://github.com/DataDog/datadog-agent/pull/5470#issuecomment-649555813
- [x] Agent status and GUI: tested manually

#### Manual test setup

In case you'd like to spin up an Agent with SNMP traps listeners enabled (eg for QA):

- Update `datadog.yaml` (QA) or `dev/dist/datadog.yaml` (local-only):

```yaml
api_key: "******"

logs_enabled: true
# logs_config:
#   # Override if you'd like to send to eg staging
#   logs_dd_url: <STAGING_LOGS_INTAKE_URL>

snmp_traps_enabled: true
snmp_traps_config:
  port: 1620
  community_strings:
    - public
```

- (Local-only) Build the Agent. I'm using:

```bash
invoke agent.build --python-home-3=~/dd/datadog-agent/venv --build-exclude=systemd,python
```

- Start the agent. For local, I'm using:

```bash
exec sudo bin/agent/agent run -c bin/agent/dist/datadog.yaml
```

- Check the `status` - should contain new info about traps. For local, I'm using:

```bash
exec sudo bin/agent/agent status -c bin/agent/dist/datadog.yaml
```

- To send a test trap:

```bash
snmptrap -v 2c -c public localhost:1620 '' NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification netSnmpExampleHeartbeatRate i 123456
```

- Trap logs should show up in Staging Live Tail with `source:snmp`.
- You can try using a wrong community string (eg `-c wrong`), send a different trap, verify the Agent behaves well if a v1 or v3 trap is directed at it, etc.

Commit:
Bootstrap new runner linked to snmp_traps.enabled

Commit:
Add log-on-receive SNMP traps receiver

prognant reviewed (commented):
I think the initial way of listening trap is ok. I think we could easily have multiple listener if needed.

I notice "_NOTE: the trap code is currently unreliable when working with snmpv3 - pull requests welcome_" on https://godoc.org/github.com/soniah/gosnmp#NewTrapListener thats probably worth digging to understand what exactly is unreliable. On the other hands gosnmp seems to support pretty much everything so that a good point.

Regarding configuration having a struct that can be directly deserialized from the `datadog.yaml` config seems the way to go.

florimondmanca commented on pkg/snmp/traps_config.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+// TrapConfig contains configuration for SNMP traps listeners.
+type TrapConfig struct {
+	Host            string
+	Port            uint16
+	Version         string
+	Timeout         int
+	Retries         int
+	Community       string
+	User            string
+	AuthKey         string
+	AuthProtocol    string
+	PrivKey         string
+	PrivProtocol    string
+	ContextEngineID string
+	ContextName     string
+}
```
YAML -> struct deserialization is nice, I think we can probably use it for most options.

I agree with the `BindHost` naming. :+1: Note: I'm taking this value from the root `bind_host` option - this is what Dogstatsd does and I think you mentioned it as well as a best practice.

Commit:
Rename Host -> BindHost, disable traps by default

Commit:
Add support for multiple listeners, refactor

Commit:
Tweak config params, update config template

Commit:
Update licenses

Commit:
Wait listeners ready, add basic server tests

florimondmanca commented on pkg/snmp/traps_config.go:
```diff
@@ -0,0 +1,94 @@
+	PrivKey         string
+	PrivProtocol    string
+	ContextEngineID string
+	ContextName     string
```
Yes you're right, the traps contain the context info.

I initially copy pasted this code from Thomas' autodiscovery PR. Providing context info in the params there makes sense because the Agent is a sender.

But I realized we're on the receiver end here, so context info is contained in trap packets (as confirmed by the GoSNMP API docs).

Timeout and retries are prob not relevant either in our case so I dropped them.

Commit:
Rename traps -> server

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Add release notes

Commit:
Refactor TrapServer.start()

Commit:
Test v2c trap reception

Commit:
Refactor, assert variables in all test cases

florimondmanca commented on pkg/snmp/traps/server_test.go:
```diff
@@ -0,0 +1,242 @@
+		require.True(t, s.Started)
+		require.Equal(t, s.NumListeners(), 1)
+
+		// FIXME: this triggers a panic during trap unmarshalling
+		// p := sendTestTrap(t, s, config)
+		// assertV3(t, p, config)
+		// assertVariables(t, p)
```
Might be a manifestation of _"[NOTE: the trap code is currently unreliable when working with snmpv3](https://godoc.org/github.com/soniah/gosnmp#GoSNMP.SendTrap)"_.

Full stack trace:

```console
--- Running unit tests:
=== RUN   TestTraps
=== RUN   TestTraps/empty
=== RUN   TestTraps/v2c-single
=== RUN   TestTraps/v3-single
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1214d96]

goroutine 32 [running]:
github.com/soniah/gosnmp.(*UsmSecurityParameters).unmarshal(0xc0001ea340, 0x4, 0xc0001c3000, 0x40, 0x1000, 0x21, 0xc0001a85c8, 0x3, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:879 +0xeb6
github.com/soniah/gosnmp.(*GoSNMP).unmarshalV3Header(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0x18, 0xc0001049a0, 0xc0001a8598, 0x3, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3.go:407 +0x4cd
github.com/soniah/gosnmp.(*GoSNMP).unmarshalHeader(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0xc0001049a0, 0xc00006ee50, 0x11c892d, 0xc0000b0220)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/marshal.go:820 +0x5b4
github.com/soniah/gosnmp.(*GoSNMP).UnmarshalTrap(0xc0001d4360, 0xc0001c3000, 0x40, 0x1000, 0x40)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:296 +0xa0
github.com/soniah/gosnmp.(*TrapListener).listenUDP(0xc0000d4f00, 0xc0001a8390, 0xf, 0x0, 0x0)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:174 +0x2cd
github.com/soniah/gosnmp.(*TrapListener).Listen(0xc0000d4f00, 0xc0001a8390, 0xf, 0x1, 0x1)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:277 +0x1c5
github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapListener).Listen(0xc000179b40)
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/listener.go:53 +0xcd
created by github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapServer).start
        /Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/server.go:71 +0x17f
FAIL    github.com/DataDog/datadog-agent/pkg/snmp/traps 0.344s
FAIL
```

Commit:
Fix and test v3 traps

florimondmanca commented:
Pushed a commit that tests V3 traps (with auth and privacy).

When testing against a live Agent I did come across a few issues/possible GoSNMP bugs, so logging stuff here for future reference…

1/ A panic occurs when Agent is not configured with privacy and client attempts to send a trap with privacy enabled, e.g.:

```console
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Packet sanity verified, we got all the bytes (191)
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed version 3
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message ID 994693058
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message max size 65507
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | parsed msg flags
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed security model 3
	2020-05-25 17:36:53 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineID ���F4�@��^
	panic: runtime error: integer divide by zero

	goroutine 212 [running]:
	github.com/soniah/gosnmp.cachedPasswordToKey(0x64eab20, 0xc000720ba0, 0x6015773, 0x3, 0x0, 0x0, 0x42d930c, 0x5b46060, 0x0, 0x0, ...)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:275 +0x485
	github.com/soniah/gosnmp.md5HMAC(0x0, 0x0, 0xc0008abae0, 0x11, 0x7dc0c4220bab3071, 0xfb344680881f0080, 0x5ecbe61440, 0x22ac2ff08236c400, 0xc0c4220bab307100)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:293 +0x9d
	github.com/soniah/gosnmp.genlocalPrivKey(0x78d0203, 0x0, 0x0, 0xc0008abae0, 0x11, 0xc000d58a70, 0x10, 0x10, 0x0, 0x0)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:380 +0x2ea
	github.com/soniah/gosnmp.(*UsmSecurityParameters).unmarshal(0xc0005b6340, 0x3, 0xc000900000, 0xbf, 0x1000, 0x30, 0xc000864940, 0x3, 0x0)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3_usm.go:853 +0x2b9
	github.com/soniah/gosnmp.(*GoSNMP).unmarshalV3Header(0xc00014d320, 0xc000900000, 0xbf, 0x1000, 0x19, 0xc000ae8ee0, 0xc00089a348, 0x3, 0x0)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3.go:407 +0x4cd
	github.com/soniah/gosnmp.(*GoSNMP).unmarshalHeader(0xc00014d320, 0xc000900000, 0xbf, 0x1000, 0xc000ae8ee0, 0xc00090ae50, 0x415ab6d, 0xc0003aa038)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/marshal.go:820 +0x5b4
	github.com/soniah/gosnmp.(*GoSNMP).UnmarshalTrap(0xc00014d320, 0xc000900000, 0xbf, 0x1000, 0xbf)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:297 +0xa0
	github.com/soniah/gosnmp.(*TrapListener).listenUDP(0xc0003a6190, 0xc0006205f0, 0xe, 0x0, 0x0)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:175 +0x2cd
	github.com/soniah/gosnmp.(*TrapListener).Listen(0xc0003a6190, 0xc0006205f0, 0xe, 0x1, 0x1)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:278 +0x1c5
	github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapListener).Listen(0xc000198a80)
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/listener.go:54 +0xcd
	created by github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapServer).start
			/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/server.go:71 +0x17f
```

2/ When client uses wrong privacy passphrase, the behavior of GoSNMP is not optimal:

- Sometimes an appropriate error is logged by GoSNMP, eg:

```console
		2020-05-25 16:40:57 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | UnmarshalTrap v3 decrypt: Error parsing SNMPV3: truncated packet
```

- Sometimes the error does not make a lot of sense to me, eg:

```console
		2020-05-25 16:42:54 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | UnmarshalTrap v3 decrypt: Error parsing SNMPV3 contextEngineID: unknown field type: 14
		2020-05-25 16:43:30 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | UnmarshalTrap v3 decrypt: Error parsing SNMPV3 contextEngineID: unknown field type: a5
```

- Sometimes (about 1 trap in 3 or 5) it crashes, somewhat similarly to https://github.com/soniah/gosnmp/issues/194, but not always at the same stage of trap processing, e.g.:

```console
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Packet sanity verified, we got all the bytes (191)
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed version 3
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message ID 2023987452
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message max size 65507
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | parsed msg flags
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed security model 3
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineID ����e����^
		2020-05-25 16:39:15 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineBoots 1
		panic: runtime error: slice bounds out of range [:-339031532695739138]

		goroutine 213 [running]:
		github.com/soniah/gosnmp.(*GoSNMP).decryptPacket(0xc00010f0e0, 0xc000707000, 0xbd, 0x1000, 0x55, 0xc000136540, 0x0, 0x0, 0xc000707000, 0x1000, ...)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3.go:442 +0xa14
		github.com/soniah/gosnmp.(*GoSNMP).UnmarshalTrap(0xc00010f0e0, 0xc000707000, 0xbf, 0x1000, 0xbf)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:310 +0x1ef
		github.com/soniah/gosnmp.(*TrapListener).listenUDP(0xc000394320, 0xc0006da380, 0xe, 0x0, 0x0)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:174 +0x2cd
		github.com/soniah/gosnmp.(*TrapListener).Listen(0xc000394320, 0xc0006da380, 0xe, 0x1, 0x1)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:277 +0x1c5
		github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapListener).Listen(0xc000186000)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/listener.go:53 +0xcd
		created by github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapServer).start
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/server.go:73 +0x17f
```

- Another example:

```console
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Packet sanity verified, we got all the bytes (191)
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed version 3
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message ID 525992801
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed message max size 65507
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | parsed msg flags
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed security model 3
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineID ���M�G���^
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineBoots 1
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authoritativeEngineTime 0
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed userName doggo
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed authenticationParameters �   ��8㋌p�5�
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed privacyParameters Ec��
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | Parsed Security Parameters. now offset=85,
		2020-05-25 16:43:43 CEST | CORE | DEBUG | (pkg/snmp/traps/config.go:32 in Printf) | UnmarshalV3Header done. [with SecurityParameters]. Header Size 79. Last 4 Bytes=[[69 99 162 212]]
		panic: runtime error: index out of range [104] with length 104

		goroutine 182 [running]:
		github.com/soniah/gosnmp.parseLength(0xc000b9e055, 0x68, 0xfab, 0x1000, 0x55)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/helper.go:572 +0x8e
		github.com/soniah/gosnmp.(*GoSNMP).decryptPacket(0xc00010afc0, 0xc000b9e000, 0xbd, 0x1000, 0x55, 0xc000a2e620, 0x0, 0x0, 0xc000b9e000, 0x1000, ...)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/v3.go:435 +0xb7
		github.com/soniah/gosnmp.(*GoSNMP).UnmarshalTrap(0xc00010afc0, 0xc000b9e000, 0xbf, 0x1000, 0xbf)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:310 +0x1ef
		github.com/soniah/gosnmp.(*TrapListener).listenUDP(0xc0001e0320, 0xc0004b0770, 0xe, 0x0, 0x0)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:174 +0x2cd
		github.com/soniah/gosnmp.(*TrapListener).Listen(0xc0001e0320, 0xc0004b0770, 0xe, 0x1, 0x1)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/vendor/github.com/soniah/gosnmp/trap.go:277 +0x1c5
		github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapListener).Listen(0xc000263460)
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/listener.go:53 +0xcd
		created by github.com/DataDog/datadog-agent/pkg/snmp/traps.(*TrapServer).start
				/Users/florimond.manca/go/src/github.com/DataDog/datadog-agent/pkg/snmp/traps/server.go:73 +0x17f
```

Commit:
Add snmp_traps_stop_timeout config option

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

florimondmanca commented:
**Notes for logs submission**

- Expose a channel on `TrapServer` to which listeners submit structured messages.
- Add a new input in `pkg/logs/inputs/traps/...`, which reads from the server's channel and dumps log messages instances to the pipeline output channel - look at the `udp`/`tcp` input implementations for inspiration.
- [Register the new input](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/agent.go#L57-L68)
- Log messages are then handled by the [existing pipeline architecture](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/agent.go#L29-L36). :sparkles:
- Expect to see some quirks relative to tagging.

Pipeline diagram:

```
(devices) --[udp:port1]--> (TrapListener 1) \
                                             (TrapServer) --[channel]--> (TrapLogsInput) --[channel]--> (Pipeline...)
(devices) --[udp:portN]--> (TrapListener N) /
```

Commit:
Add integration test for noauth-nopriv and auth-nopriv, refactor config building

Commit:
Add copyright notices

Commit:
Add TrapListenerConfig unit tests

Commit:
Refactor tests, add mixed concurrent listeners test

Commit:
Add logs forwarding plumbing

florimondmanca commented:
Making good progress on logs forwarding:

<img width="1418" alt="Screenshot 2020-05-27 at 15 57 35" src="https://user-images.githubusercontent.com/15911462/83037984-be55d700-a03c-11ea-8d33-6de2b7c9e247.png">

Commit:
Format trap logs (as JSON)

Commit:
Add more fields to trap logs

Commit:
Fix server_test

Commit:
Align config naming on SNMP autodiscovery, make port required, handle and test more config error cases

Commit:
Tweak config template presentation

Commit:
Cleanup TODOs

Commit:
Document bind_host option

Commit:
Drop packets with wrong community string

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Fix nil pointer panic in case of test failure

prognant commented on pkg/config/config_template.yaml:
```diff
@@ -1588,3 +1588,79 @@ api_key:
+
+    ## @param  version - string - optional
+    ## The SNMP version to use.
+    ## One of: "1" (SNMPv1), "2c" (SNMPv2c), "3" (SNMPv3).
```
Do we really want SNMP 1 (I would say no, but who knows, some customer may have hardware from the 90s)

florimondmanca replied:
It's cheap enough to support it (built into GoSNMP w/o any extra work), but I don't have any background yet on whether this is a required feature for customers.

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Test wrong user, authkey or privkey

Commit:
Refactor server tests

Commit:
Add more bad credentials test cases

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Drop v3

prognant commented on pkg/logs/input/traps/tailer.go:
```diff
@@ -0,0 +1,56 @@
+			log.Errorf("failed to format packet: %s", err)
+			continue
+		}
+		t.outputChan <- message.NewMessageWithSource(content, message.StatusInfo, t.source)
```
We may add some tagging logic here (switch to `message.NewMessage(content []byte, origin *Origin, status string)` and populate `origin` with tags like the sender address, the snmp version), it could be done later as it does not imply any refactoring.

prognant commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+var (
+	outputChannelSize = 100
```
Maybe that one could become a (hidden) configuration setting.

truthbk replied:
This could indeed be a nice knob, it's no huge deal if we don't add it yet as it can be easily included in the future, but nice-to-have.

florimondmanca replied:
Yes, eventually we could easily make this configurable. For now I'll move this with the other constants you suggested creating. :-)

prognant commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,170 @@
+
+	select {
+	case <-allReady:
+	case err := <-readyErrors:
+		close(readyErrors)
+		return err
+	}
```
If a single listener fails the entire process including working listener is considered as failed. That prevents potentially working listener to be started. IMHO we should support starting valid listener even if broken ones are included in the configuration.

prognant commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,170 @@
+		stopTimeout: stopTimeout,
+	}
+
+	err = s.start()
```
Here if we have a single listener that fails, all listener, including valid ones, will be ignored.

florimondmanca replied:
I think this is now fixed! Invalid listeners are just ignored, allowing valid ones to start.

prognant commented on cmd/agent/app/run.go:
```diff
@@ -278,6 +279,14 @@ func StartAgent() error {
 	log.Debugf("statsd started")
 
+	// Start SNMP-related services
+	if traps.IsEnabled() {
```
The trap feature relies on logs, to avoid loopholes like traps enabled but logs disabled, we should check whether logs are enabled in `traps.IsEnabled()` and log a warning if we have logs disabled.

florimondmanca replied:
Good point! I went for skipping starting the trap server entirely if logs are disabled. (Otherwise it'd be useless work for the Agent and I'm not sure we wouldn't run into issues w/ the server output channel filling up over time etc…)

prognant commented on releasenotes/notes/snmp-traps-4f0e9ba2a6247322.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add support for receiving and processing SNMP traps, and forwarding them to Datadog.
```
We should mentioned that traps are forwarded as logs.

prognant reviewed (dismissed):
Could you please also checked that `agent status` reports new source correctly in logs. 
And I think we should expose expvar and relay them as well into `agent status` under a new Trap section (see pkg/status/status.go for additional details).

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Address feedback in pkg/logs/input/traps

Commit:
Skip trap server if logs disabled, update release note, add tags to logs

florimondmanca commented on pkg/snmp/traps/config.go:
```diff
@@ -0,0 +1,76 @@
+		return nil, err
+	}
+
+	if c.Community == "" {
+		return nil, errors.New("`community` is required")
+	}
```
Is this what you are thinking about?

```yaml
snmp_traps_enabled: true

snmp_traps_listeners:
  - version: '2'
    community:
      - public
      - dev123
      - internalABC
    port: 162
```

florimondmanca replied:
I pushed a PR that implements something like the snippet above^ and update the RFC.

Commit:
Add multi-community support, drop v1 support

florimondmanca commented:
@prognant Thanks for reviewing!

I still need to address the "allow valid listeners to start and skip invalid ones" part, but I think I've addressed the rest of your comments.

Note that I decided to drop SNMPv1 from this PR for now; doing a bit more manual testing it turned out there would be some more work required (eg because the PDU format is a bit different and there might be some extra required options). So I moved it to "phase 2" as defined by the RFC.

Commit:
Update config_template.yaml

Commit:
Move format.go to snmp/traps, remove unused version checking

prognant reviewed (commented):
Thanks for the update !

prognant commented:
Juste a side note, for future work I think we should plan at some point to add some benchmark in tests like [this](https://github.com/DataDog/datadog-agent/blob/master/pkg/metrics/series_benchmark_test.go) to assess the whole trap ingestion logic performance.

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Refactor test code, update listener config, add format tests

Commit:
Allow valid listeners to start even if others are invalid

florimondmanca commented:
New traps source is correctly shown in the `agent status`:

```console
==========
Logs Agent
==========

    ...

    BytesSent: 0
    EncodedBytesSent: 28
    LogsProcessed: 0
    LogsSent: 0

  snmp_traps
  ----------
    Type: traps
    Status: OK
```

Commit:
Tweak presentation of snmp_traps logs source

AlexandreYang commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,174 @@
+
+	output := make(OutputChannel, outputChannelSize)
+	listeners := make([]TrapListener, 0, len(configs))
+	running := make([]bool, 0, len(configs))
```
Any reason `running` is not a boolean field on `TrapListener` ?

florimondmanca replied:
This is initially what I did: `TrapListener` would manage a `.Running` flag that it sets and unsets in `.Listen()`.

But I ran into race conditions in tests. Because `.Listen()` is run in a goroutine, if the rest of operations are sync (like in a test where I just inspect the number of running listeners) then the flag is only set after those operations are done.

So even if `NewTrapServer()` returned a started server, the goroutines hadn't started and the flags weren't set. So instead since the server is the one managing the lifecycle of listeners, I put the "keep track of registered/running/failed listeners" logic there.

florimondmanca replied:
FYI, pushed the commit that flips how this state is tracked (running listeners -> failed listeners). Let me know if how this looks.

AlexandreYang reviewed (changes requested):
Left few general comments.

florimondmanca commented on pkg/logs/config/config.go:
```diff
@@ -57,6 +61,15 @@ func DefaultSources() []*LogSource {
 
+	if traps.IsEnabled() {
+		source := NewLogSource(SnmpTrapsCollect, &LogsConfig{
+			Type:    "traps",
```
Fixed as of the latest state of this PR. :-)

florimondmanca commented on pkg/snmp/traps/config.go:
```diff
@@ -0,0 +1,61 @@
+	Version          uint16   `mapstructure:"version" yaml:"version"`
+	Port             uint16   `mapstructure:"port" yaml:"port"`
+	CommunityStrings []string `mapstructure:"community_strings" yaml:"community_strings"`
+	BindHost         string   `mapstructure:"bind_host" yaml:"bind_host"`
```
There's a comment above the `struct` declaration that clarifies that these are needed to facilitate tests (I generate the YAML from structs to make sure the conf parsing works, which `snmp.go` doesn't test).

See: https://github.com/DataDog/datadog-agent/pull/5470/files#diff-2defe81bba4f82bb622e334cdc9a85e6R154-R160

AlexandreYang replied:
I see.

However, a slightly unrelated: maybe you can use text in tests instead of Marshaling objects if that's possible (I didn't had time to review the test part yet but will do soon)

florimondmanca replied:
Yes, it's possible to use text, but we end up having to essentially re-write a struct-to-YAML serializer (or we have to duplicate the config and the YAML file). This way with the `yaml:` field keeps the tests cleaner.

florimondmanca commented on pkg/snmp/traps/format.go:
```diff
@@ -0,0 +1,145 @@
+// GetTags returns a list of tags associated to an SNMP trap packet.
+func GetTags(p *SnmpPacket) []string {
+	return []string{
+		fmt.Sprintf("snmp_version:2"),
```
Sure, makes sense, I can bring back my `GetVersion(p *SnmpPacket)` helper I had before.

florimondmanca commented on pkg/snmp/traps/format.go:
```diff
@@ -0,0 +1,145 @@
+		return 0, fmt.Errorf("expected OID %s or %s, got %s", sysUpTime, sysUpTimeInstance, v.Name)
+	}
+
+	value, ok := v.Value.(uint32)
+	if !ok {
+		return 0, fmt.Errorf("expected uptime to be uint32 (got %v of type %T)", v.Value, v.Value)
+	}
+
+	// sysUpTimeInstance is given in hundreds of a second, convert it to seconds.
+	value = value / 100
+
+	return value, nil
```
Ah you're right, I hadn't realized that this `value` is an `uint32` then `value / 100` means we lose the decimal places.

Converting to ms by doing `value *= 10` is definitely better. :+1:

AlexandreYang replied:
It seems that we don’t we don’t make scale change for Timeticks for sysUpTimeInstance in snmp integration. I think we should be consistent and not change it here either.

Not sure what are the impact of using hundredth of seconds from UI perspective.

AlexandreYang replied:
Note, there are some discussion about adding `centisecond` for metrics unit.

cc @naaltman

florimondmanca replied:
Thanks for the heads up. For now I just went with the value unchanged.

Note that the `uptime` here is only indicative. It is not added for metrics purposes, and it is not used as the log message timestamp either (a timestamp is added automatically by the Agent: https://github.com/DataDog/datadog-agent/blob/1c76b8381a195a0b0f629011a6225e936fe1d37a/pkg/logs/processor/json.go#L39).

florimondmanca commented on pkg/snmp/traps/format.go:
```diff
@@ -0,0 +1,145 @@
+		return "string"
+	case gosnmp.ObjectIdentifier:
+		return "oid"
+	case gosnmp.Counter32, gosnmp.Counter64:
+		return "counter"
+	case gosnmp.Gauge32:
+		return "gauge"
```
I'm not sure I'll keep this in the final log format, but you're right, we should keep a more straight-forward mapping.

Commit:
Add expvars and status section, refactor failed listeners counting

florimondmanca commented:
Added basic expvars and a new template in the `agent status` (only shown if traps are enabled):

```console
$ agent status

...

=========
SNMP Traps
==========
  Listeners: 1
  Listeners Start Failed: 0
  Packets: 2
  Packets Auth Errors: 1
```

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Simplify, polishing

Commit:
Address more feedback

Commit:
Add log tailer tests, simplify format tests

florimondmanca commented on pkg/config/config_template.yaml:
```diff
@@ -1745,3 +1745,51 @@ api_key:
+{{- if .SNMP }}
+
+########################
+## SNMP Configuration ##
```
I actually think the `snmp_listener` section is misplaced, and it should be moved here once this PR is merged. (It's currently in the "Advanced section" alongside options such as confd_path`, `secret_backend_output_max_size`, etc, which is not ideal.)

florimondmanca committed:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Drop outdated pipeline diagram

florimondmanca commented:
Notes for Agent Core review:

The updated pipeline diagram looks like this:

```console
       (Outside world) | (Agent)
                       |
(devices) --[udp:port]-|-> (TrapServer) --[channel]--> (TrapLogsInput) --[channel]--> (Pipeline...)
                       |
```

- Does the general approach look good?
- Is there anything missing that I may not be aware of? (For example, Pierre mentioned updating the `status` output, which I would not have thought of myself.)
- Can you pay attention to the code in `config.go` (config parsing), in case I'm missing any gotcha?
- Can you pay attention to the code in `format.go` (the formatting of log messages)? I'm not super-satisfied with how the manual switches on type and casting look. There may be more "Go-idiomatic" ways to format and cast data. If so, let me know!
- Can you pay attention to the handling of channels? Any possible gotchas relative to backpressure? (For example, in case of high number of incoming traps I assume that putting a parsed packet in the `(TrapServer) -> (TrapLogsInput)` packets channel will block until the logs input has processed and sent the packet to the logs pipeline. Is this a good approach?)

florimondmanca committed:
Merge branch 'master' into florimondmanca/snmp-traps

AlexandreYang commented on pkg/snmp/traps/format.go:
```diff
@@ -0,0 +1,157 @@
+
+func parseSysUpTime(v gosnmp.SnmpPDU) (uint32, error) {
+	name := normalizeOID(v.Name)
+	if name != sysUpTimeOID && name != sysUpTimeInstanceOID {
```
Is `sysUpTimeOID` a real use case without `.0` ? The RFC seems to indicate that the OID is an instance OID with `.0`.

Same comment for `snmpTrapOID`.

florimondmanca replied:
I think given the mess that is SNMP, it's possible that some devices send it without the `.0`? I'm just being overly cautious here. 😅

AlexandreYang replied:
> I think given the mess that is SNMP, it's possible that some devices send it without the .0?

I can't say for sure, but probably unlikely, if it happen, it's a bug on the device. 

IMHO it's better to start with only support the instance OID (with `.0`) and add fallback if there are good reasons to support it later.

Also, we assume here, the OID is an instance OID with `.0`: https://github.com/DataDog/integrations-core/blob/e8eb6484a7aea40f5919929e02608cbe4babaacf/snmp/datadog_checks/snmp/config.py#L299

florimondmanca replied:
Okay, so it looks like there's actually no such thing as `snmpTrapOIDInstance` - it's really just `snmpTrapOID` without a `.0`. But you're right, it should be `sysUptimeInstance` and never `sysUptime`. Fixed!

florimondmanca replied:
Okay, so it looks like there's actually no such thing as `snmpTrapOIDInstance` - it's really just `snmpTrapOID` without a `.0`. But you're right, it should be `sysUptimeInstance` and never `sysUptime`. Fixed!

AlexandreYang commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,147 @@
+
+// NewTrapServer configures and returns a running SNMP traps server.
+func NewTrapServer() (*TrapServer, error) {
+	c, err := ReadConfig()
```
```suggestion
	config, err := ReadConfig()
```

(minor) c => config, improves readability. There are multiple locations where single letter is used, which is concise, but makes the code less readable.

IMO, we should avoid single letter vars and acronyms in general.

florimondmanca replied:
From what I've seen single-letter variables isn't uncommon in the Golang world, but I agree. :-) I'll change them when it's obvious we can avoid them.

AlexandreYang reviewed (commented):
My review is mainly focus on things related to `snmp` itself.

truthbk commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,147 @@
+			trapsPacketsAuthErrors.Add(1)
+			return
+		}
+		log.Debugf("New valid packet received from %s on listener %s", u.String(), c.Addr())
```
Question: might this be too verbose even as debug mode? Trap volume can potentially be quite high... we can see from the counters if there are incoming traps being processed.

florimondmanca replied:
Good point, I'm removing this log line.

truthbk replied:
Or make it trace-level, whatever you think is best.

truthbk commented on pkg/snmp/traps/server.go:
```diff
@@ -0,0 +1,147 @@
+		}
+	}()
+
+	select {
```
Though the code is pretty self explanatory we might want to add a comment with this: https://godoc.org/github.com/soniah/gosnmp#TrapListener.Listening.

truthbk reviewed (commented):
A few tiny comments, but this is very nicely done. 🙇

florimondmanca commented on pkg/logs/config/config.go:
```diff
@@ -57,6 +61,16 @@ func DefaultSources() []*LogSource {
 		sources = append(sources, source)
 	}
 
+	if traps.IsEnabled() && traps.IsRunning() {
```
The sequence of operations is as follows:

- Trap server is setup in `run.go`
  - Trap server starts, we wait for it to be ready to accept traps, then set `server` (-> `IsRunning()` returns `true`).
- Logs Agent is setup in `run.go`
  - This code is run, and this condition is true.

The key aspect is that the logs setup needs to be done _after_ the traps setup. I don't think it's super clear in the code right now, I should probably add a comment in `run.go` about this…?

florimondmanca commented on pkg/snmp/traps/config.go:
```diff
@@ -0,0 +1,72 @@
+func (c *Config) BuildV2Params() *gosnmp.GoSNMP {
+	return &gosnmp.GoSNMP{
+		Port:      c.Port,
+		Transport: "udp",
```
In the future we might, but for this "MVP" first step we're only supporting UDP (which is by far the most common transport for traps AFAIK).

florimondmanca commented on pkg/status/templates/snmp-traps.tmpl:
```diff
@@ -0,0 +1,13 @@
+SNMP Traps
+==========
+{{- if .error }}
+  Error: {{.error}}
```
There is a `PacketsAuthErrors` expvar already, and this `Error` field is mostly for "startup failures" (eg "failed to bind socket"). Do you think we need something else?

florimondmanca commented on pkg/config/config.go:
```diff
@@ -407,6 +407,11 @@ func InitConfig(config Config) {
+	config.BindEnvAndSetDefault("snmp_traps_enabled", false)
+	config.BindEnvAndSetDefault("snmp_traps_config.port", 162)
+	config.BindEnvAndSetDefault("snmp_traps_config.community_strings", []string{})
+	config.BindEnvAndSetDefault("snmp_traps_config.bind_host", "localhost")
```
For the sake of this PR I'm going for the absolute minimum config we need, not thinking about the future v3 support (which might arrive in a while). When we do add v3, we can either break things (this is marked as *EXPERIMENTAL* after all), or find a way to keep backwards compat.

Eg a possible configuration that would limit the compat code would be:

```yaml
snmp_traps_config:
  v2:
    port: 162
    community_strings:
      - public0
      - public1
      - public2
  v3:
    - port: 163
      user: doggo
      authProtocol: MD5
      authKey: doggopass
      privProtocol: AES
      privKey: doggokey
    - ...
```

Essentially we'd move the current config to a `v2` field, and adds a `v3` field with v3-specific config (multiple listeners). We could fallback to applying parsing of `TrapV2Config` if none of `v2` or `v3` is found, and raise a deprecation warning.

Another possibility would be to just do a list of listeners with a `version`, as before. But this brings ambiguity in terms of "in what case would I want multiple v2 listeners" (to which I'd answer "none", but I'm not 100% certain, so leaving the door open…).

AlexandreYang replied:
Yes, right, it's experimental. Sure, let's keep it simple for now 👍

florimondmanca commented on pkg/logs/config/config.go:
```diff
@@ -13,12 +13,16 @@ import (
 const ContainerCollectAll = "container_collect_all"
 
+// SnmpTraps is the name of the integration that collects logs from SNMP traps received by the Agent
+const SnmpTraps = "snmp_traps"
```
@truthbk Do you have any insight on this? My guess for the Docker case is that obviously both would be set to "docker" so the tests use the constant interchangeably (?)…

(@AlexandreYang Note that the line you're highlighting here does not define the log source name nor the log type - it's the log integration name. If anything we might want to rename it to `snmp_trap_collect_all` to match `container_collect_all`, I don't have a strong opinion about this…)

AlexandreYang replied:
> (@AlexandreYang Note that the line you're highlighting here does not define the log source name nor the log type - it's the log integration name. If anything we might want to rename it to snmp_trap_collect_all to match container_collect_all, I don't have a strong opinion about this…)

Yes, I was referring to the integration name, which is (if I'm not mistaken) save on `NewLogSource.name`.

florimondmanca replied:
I don't think there's a rule. If you look at the `ContainerCollectAll` log integration above, it uses `container_collect_all` as the log source name, yet the type is `DockerType`. So I'll stick to the current `snmp_traps` integration name for now…

truthbk replied:
Sorry, I missed this notification last week, indeed, I believe the current LogSource definition for the log source is sendible.

florimondmanca commented on pkg/snmp/traps/format.go:
```diff
@@ -0,0 +1,157 @@
+func GetTags(p *SnmpPacket) []string {
+	tags := []string{
+		fmt.Sprintf("community:%s", p.Content.Community),
+		fmt.Sprintf("device_ip:%s", p.Addr.IP.String()),
+		fmt.Sprintf("device_port:%d", p.Addr.Port),
```
Good call!

Commit:
Address feedback

truthbk reviewed (commented):
There's a formatting issue preventing the unit tests from running, but otherwise this is looking good to me! 👍

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -1816,3 +1816,47 @@ api_key:
+# snmp_traps_enabled: false
+
+## @param snmp_traps_config - custom object - optional
+## This section configures SNMP traps collection. Traps are forwarded as logs to the Datadog Platform.
```
```suggestion
## This section configures SNMP traps collection. Traps are forwarded as logs to Datadog.
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -1816,3 +1816,47 @@ api_key:
+
+## @param snmp_traps_config - custom object - optional
+## This section configures SNMP traps collection. Traps are forwarded as logs to the Datadog Platform.
+## NOTE: this feature is currently **EXPERIMENTAL**. Both behavior and configuration options may
```
```suggestion
## NOTE: This feature is currently **EXPERIMENTAL**. Both behavior and configuration options may
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -1816,3 +1816,47 @@ api_key:
+## @param snmp_traps_config - custom object - optional
+## This section configures SNMP traps collection. Traps are forwarded as logs to the Datadog Platform.
+## NOTE: this feature is currently **EXPERIMENTAL**. Both behavior and configuration options may
+## change in the future. Only SNMPv2 is supported for now.
```
```suggestion
## change in the future. Only SNMPv2 is supported.
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -1816,3 +1816,47 @@ api_key:
+
+  ## @param community_strings - list of strings - required
+  ## A list of known SNMPv2 community strings that devices can use to send traps to the Agent.
+  ## Traps with an unknown community string will be ignored.
```
```suggestion
  ## Traps with an unknown community string are ignored.
```

ruthnaebeck reviewed (changes requested):
Docs review

Commit:
Run gofmt

florimondmanca committed:
Apply suggestions from docs review

Co-authored-by: ruthnaebeck <19349244+ruthnaebeck@users.noreply.github.com>

ruthnaebeck reviewed (approved):
👍 for docs

AlexandreYang commented:
@florimondmanca Looks good for me overall. Two open discussion left on my side: 
- https://github.com/DataDog/datadog-agent/pull/5470/files#r462928870
- https://github.com/DataDog/datadog-agent/pull/5470#discussion_r462850509

Commit:
Address feedback, review tests

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

florimondmanca commented:
@AlexandreYang Addressed your comments :+1:

AlexandreYang commented:
Looks good, thx for the changes.

truthbk reviewed (commented):
Almost there @florimondmanca, just a couple of things on the logs side for consistency and cleanliness, and a tiny thing with the logging (that is actually my fault).

AlexandreYang commented on pkg/logs/config/integration_config.go:
```diff
@@ -18,6 +18,7 @@ const (
 	DockerType       = "docker"
 	JournaldType     = "journald"
 	WindowsEventType = "windows_event"
+	SnmpTrapType     = "snmp_trap"
```
I would voucher for singular `snmp_trap` and change elsewhere to be consistent:
- snmp_traps_config=> snmp_trap_config
- snmp_traps_enabled=> snmp_trap_enabled
- expvar.NewMap("snmp_traps")=>expvar.NewMap("snmp_trap")

It's itself consistent with existing `windows_event` and `snmp_listener` that are singular.

florimondmanca replied:
@AlexandreYang I would go with Jaime's suggestion here. I understand your point but in my mind we talk about "collecting _(multiple)_ trapS " in general it's more natural to just have `snmp_traps` elsewhere (as opposed to `snmp_trap` elsewhere). I don't know why `windows_event` was decided to be singular. But `snmp_listener` makes sense since there's really only one SNMP autodiscovery per Agent (although it runs multiple workers).

Commit:
Merge branch 'master' into florimondmanca/snmp-traps

Commit:
Address feedback

truthbk reviewed (approved):
Yes! 🍰 🥇

AlexandreYang commented:
cc @florimondmanca 

# QA Feedbacks:

1/ PR  snmpTrapOID should be an instance OID by using instance OID

Fix: https://github.com/DataDog/datadog-agent/pull/6208

Details: 
Running

```
snmptrap -v 2c -c public localhost:1162 '' NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification netSnmpExampleHeartbeatRate i 123456

snmptrap --version
NET-SNMP version: 5.7.3

```

Getting this error:

```
2020-08-12 16:20:58 UTC | CORE | ERROR | (pkg/logs/input/traps/tailer.go:54 in run) | failed to format packet: expected OID 1.3.6.1.6.3.1.1.4.1, got 1.3.6.1.6.3.1.1.4.1.0
```

3/ PR  timeout config is missing from config.go

5/ PR  It seems that `snmp_device_port` port tag might vary for each trap received and lead to a lot of contexts. 

```
fmt.Sprintf("snmp_device_port:%d", packet.Addr.Port),
```

At least when using:

```
snmptrap -v 2c -c abc 10.0.2.15:1162 '' NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification netSnmpExampleHeartbeatRate i 12345
```

=> TODO now: Remove `snmp_device_port` tag

7/ I wasn't able to test timeout since Traps stops less than a second

=> Alex QA: Add sleep to https://github.com/DataDog/datadog-agent/blob/9e2ad3b72da3e98d5b91484f70bbaab1060a720f/pkg/snmp/traps/server.go#L135

=> Improvement: make config.StopTimeout testable.


# DO LATER

2/ `variables` items are "string" not json. @florimondmanca_dd is that expected ?

https://trello-attachments.s3.amazonaws.com/5f314267494c71120f72e42b/600x279/ca5cdc2df037cd774925d8649be145f6/image.png

=> TODO: example: use a map of variables using OID as key (for example). Make sure users can query the OID and value using logs query engine.


4/ [can be addressed later] Defaults are already set in `config.go`, no need to have constants in code like: 

- `defaultStopTimeout`
- `defaultPort`


6/ SNMP types

- decimal strings seems not supported

```
http://www.net-snmp.org/docs/man/snmptrap.html
```
=> Support for Decimals is FR

- UNSIGNED reported as gauge32

```
snmptrap -v 2c -c abc 10.0.2.15:1162 '' NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification         1.3.6.1.2.1.74.1.2.2.1 u 123
```




# QA details

### QA case: On invalid community string:

OK

```
Invalid credentials from 127.0.0.1:46063 on listener localhost:1162, dropping packet
```

### QA case: multiple community string: abc, public

OK


### QA case: snmp_traps_enabled:false/true

OK

`snmp_traps_enabled:false` disables snmp traps as expected

### QA case: invalid bind_host

OK

```
2020-08-13 09:23:56 UTC | CORE | INFO | (pkg/snmp/traps/server.go:107 in func2) | Start listening for traps on 1.1.1.1:1162
2020-08-13 09:23:56 UTC | CORE | ERROR | (cmd/agent/app/run.go:288 in StartAgent) | Failed to start snmp-traps server: listen udp 1.1.1.1:1162: bind: cannot assign requested address
```

### QA case: valid bind_host different than default

OK

```
2020-08-13 09:26:34 UTC | CORE | DEBUG | (pkg/snmp/traps/server.go:98 in func1) | Packet received from 10.0.2.15:59401 on listener 10.0.2.15:1162
```

### QA case: stop correctly when agent stops

OK

```
2020-08-13 09:36:42 UTC | CORE | INFO | (pkg/snmp/traps/server.go:134 in func1) | Stop listening on 10.0.2.15:1162
```

### QA case: tags correctly send to logs backend

```
TAGS
env:alex-vagrant-dev
service:snmp
snmp_device:127.0.0.1
snmp_device_port:57923
snmp_version:2
source:snmp
```

### QA case: value types

- UNSIGNED INT, reported as gauge32
- int reported as in, OK
- string OK
- counter32 OK
- counter64 NOT TESTED (not available for snmptrap command http://www.net-snmp.org/docs/man/snmptrap.html)
